basic/radix-name-service/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+use scrypto::prelude::*;
+use sha2::{Digest, Sha256};
+
+#[derive(NftData)]
+struct DomainName {
+    #[scrypto(mutable)]
+    address: Address,
+    #[scrypto(mutable)]
+    last_valid_epoch: u64,
+    #[scrypto(mutable)]
+    deposit_amount: Decimal,
+}
+
+// Assuming an average epoch duration of 35 mninutes, 15k epochs very roughly fit into one year
+// This is just a very rough estimate, of course
+const EPOCHS_PER_YEAR: u64 = 15_000;
+
+blueprint! {
+
+    struct RadixNameService {
+        admin_badge: ResourceDef,
+        minter: Vault,
+        name_resource: ResourceDef,
+        deposits: Vault,
+        fees: Vault,
+        deposit_per_year: Decimal,
+        fee_address_update: Decimal,
+        fee_renewal_per_year: Decimal,
+    }
+
+    impl RadixNameService {
+        /// Creates a new RNS instance
+        pub fn new() -> (Component, Bucket) {
+            let admin_badge =
+                ResourceBuilder::new_fungible(DIVISIBILITY_NONE).initial_supply_fungible(1);
+
+            let minter =
+                ResourceBuilder::new_fungible(DIVISIBILITY_NONE).initial_supply_fungible(1);
+
+            let name_resource = ResourceBuilder::new_non_fungible()
+                .metadata("name", "DomainName")
+                .flags(MINTABLE | BURNABLE | INDIVIDUAL_METADATA_MUTABLE | RECALLABLE)
+                .badge(minter.resource_def(), ALL_PERMISSIONS)
+                .no_initial_supply();
+
+            let component = RadixNameService {
+                admin_badge: admin_badge.resource_def(),
+                minter: Vault::with_bucket(minter),
+                name_resource,
+                deposits: Vault::new(RADIX_TOKEN),
+                fees: Vault::new(RADIX_TOKEN),
+                deposit_per_year: 50.into(),
+                fee_address_update: 10.into(),
+                fee_renewal_per_year: 25.into(),
+            }
+            .instantiate();
+
+            (component, admin_badge)
+        }
+
+        /// Lookup the address for a given `name`.
+        /// Panics if that name is not registered.
+        pub fn lookup_address(&self, name: String) -> Address {
+            let hash = Self::hash_name(name);
+            let name_data: DomainName = self.name_resource.get_nft_data(hash);
```

Comment:
It will be even better for composability when we change `get_nft_data` to return an Option instead of panicking